File: src/storage/state-store.ts
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";
import { chmod, mkdir, readFile, rename, rm, writeFile } from "node:fs/promises";
import os from "node:os";
import path from "node:path";

import type { StoredConfig, StoredCredentials, StoredSession } from "../types.js";

interface EncryptedPayload {
  iv: string;
  tag: string;
  ciphertext: string;
}

const DEFAULT_DIRNAME = ".docmost-local-mcp";

export class StateStore {
  readonly baseDir: string;
  private readonly configPath: string;
  private readonly sessionPath: string;
  private readonly credentialsPath: string;
  private readonly keyPath: string;

  constructor(baseDir = path.join(os.homedir(), DEFAULT_DIRNAME)) {
    this.baseDir = baseDir;
    this.configPath = path.join(this.baseDir, "config.json");
    this.sessionPath = path.join(this.baseDir, "session.json");
    this.credentialsPath = path.join(this.baseDir, "credentials.enc.json");
    this.keyPath = path.join(this.baseDir, "credentials.key");
  }

  async readConfig(): Promise<StoredConfig | null> {
    return this.readJsonFile<StoredConfig>(this.configPath);
  }

  async writeConfig(config: StoredConfig): Promise<void> {
    await this.writeJsonFile(this.configPath, config);
  }

  async readSession(): Promise<StoredSession | null> {
    return this.readJsonFile<StoredSession>(this.sessionPath);
  }

  async writeSession(session: StoredSession): Promise<void> {
    await this.writeJsonFile(this.sessionPath, session);
  }

  async clearSession(): Promise<void> {
    await rm(this.sessionPath, { force: true });
  }

  async readCredentials(): Promise<StoredCredentials | null> {
    const encrypted = await this.readJsonFile<EncryptedPayload>(this.credentialsPath);
    if (!encrypted) {
      return null;
    }

    const key = await this.getOrCreateEncryptionKey();
    const plaintext = this.decryptString(encrypted, key);
    return JSON.parse(plaintext) as StoredCredentials;
  }

  async writeCredentials(credentials: StoredCredentials): Promise<void> {
    const key = await this.getOrCreateEncryptionKey();
    const encrypted = this.encryptString(JSON.stringify(credentials), key);
    await this.writeJsonFile(this.credentialsPath, encrypted);
  }

  private async ensureBaseDir(): Promise<void> {
    await mkdir(this.baseDir, { recursive: true, mode: 0o700 });
    await chmod(this.baseDir, 0o700);
  }

  private async readJsonFile<T>(filePath: string): Promise<T | null> {
    try {
      const contents = await readFile(filePath, "utf8");
      return JSON.parse(contents) as T;
    } catch (error) {
      if (isMissingFile(error)) {
        return null;
      }

      throw error;
    }
  }

  private async writeJsonFile(filePath: string, value: unknown): Promise<void> {
    await this.ensureBaseDir();

    const tempPath = `${filePath}.tmp`;
    const contents = `${JSON.stringify(value, null, 2)}\n`;
    await writeFile(tempPath, contents, { encoding: "utf8", mode: 0o600 });
    await chmod(tempPath, 0o600);
    await rename(tempPath, filePath);
  }

  private async getOrCreateEncryptionKey(): Promise<Buffer> {
    await this.ensureBaseDir();

    try {
      const stored = await readFile(this.keyPath, "utf8");
      return Buffer.from(stored.trim(), "base64");
    } catch (error) {
      if (!isMissingFile(error)) {
        throw error;
      }
    }

    const key = randomBytes(32);
    await writeFile(this.keyPath, key.toString("base64"), { encoding: "utf8", mode: 0o600 });
    await chmod(this.keyPath, 0o600);
    return key;
  }

  private encryptString(plaintext: string, key: Buffer): EncryptedPayload {
    const iv = randomBytes(12);
    const cipher = createCipheriv("aes-256-gcm", key, iv);
    const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
    const tag = cipher.getAuthTag();

    return {
      iv: iv.toString("base64"),
      tag: tag.toString("base64"),
      ciphertext: ciphertext.toString("base64"),
    };
  }

  private decryptString(payload: EncryptedPayload, key: Buffer): string {
    const decipher = createDecipheriv(
      "aes-256-gcm",
      key,
      Buffer.from(payload.iv, "base64"),
    );
    decipher.setAuthTag(Buffer.from(payload.tag, "base64"));

    const plaintext = Buffer.concat([
      decipher.update(Buffer.from(payload.ciphertext, "base64")),
      decipher.final(),
    ]);

    return plaintext.toString("utf8");
  }
}

function isMissingFile(error: unknown): error is NodeJS.ErrnoException {
  return (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    error.code === "ENOENT"
  );
}

File: src/storage/state_store.rs
use std::path::{Path, PathBuf};

use aes_gcm::{
    Aes256Gcm, KeyInit, Nonce,
    aead::{Aead, OsRng, rand_core::RngCore},
};
use anyhow::{Context, Result};
use base64::{Engine as _, engine::general_purpose::STANDARD};
use serde::{Serialize, de::DeserializeOwned};
use tokio::fs;

use crate::{
    storage::keyring_store::KeyringStore,
    types::{StoredConfig, StoredCredentials, StoredSession},
};

const DEFAULT_DIRNAME: &str = ".docmost-local-mcp";

#[derive(Debug, Clone)]
pub struct StateStore {
    pub base_dir: PathBuf,
    config_path: PathBuf,
    session_path: PathBuf,
    credentials_path: PathBuf,
    key_path: PathBuf,
    keyring: KeyringStore,
}

#[derive(Debug, Serialize, serde::Deserialize)]
struct EncryptedPayload {
    iv: String,
    tag: String,
    ciphertext: String,
}

impl StateStore {
    pub fn new(base_dir: Option<PathBuf>) -> Result<Self> {
        let base_dir = match base_dir {
            Some(base_dir) => base_dir,
            None => dirs::home_dir()
                .context("Unable to determine the home directory")?
                .join(DEFAULT_DIRNAME),
        };

        Ok(Self {
            config_path: base_dir.join("config.json"),
            session_path: base_dir.join("session.json"),
            credentials_path: base_dir.join("credentials.enc.json"),
            key_path: base_dir.join("credentials.key"),
            base_dir,
            keyring: KeyringStore,
        })
    }

    pub async fn read_config(&self) -> Result<Option<StoredConfig>> {
        self.read_json_file(&self.config_path).await
    }

    pub async fn write_config(&self, config: &StoredConfig) -> Result<()> {
        self.write_json_file(&self.config_path, config).await
    }

    pub async fn read_session(&self) -> Result<Option<StoredSession>> {
        self.read_json_file(&self.session_path).await
    }

    pub async fn write_session(&self, session: &StoredSession) -> Result<()> {
        self.write_json_file(&self.session_path, session).await
    }

    pub async fn clear_session(&self) -> Result<()> {
        match fs::remove_file(&self.session_path).await {
            Ok(()) => Ok(()),
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(()),
            Err(error) => Err(error).context("Failed to clear saved session"),
        }
    }

    pub async fn read_credentials(&self) -> Result<Option<StoredCredentials>> {
        if let Some(credentials) = self.keyring.read_credentials()? {
            return Ok(Some(credentials));
        }

        let Some(payload) = self
            .read_json_file::<EncryptedPayload>(&self.credentials_path)
            .await?
        else {
            return Ok(None);
        };

        let key = self.get_or_create_encryption_key().await?;
        let plaintext = decrypt_string(&payload, &key)?;
        let credentials =
            serde_json::from_str(&plaintext).context("Failed to parse decrypted credentials")?;
        Ok(Some(credentials))
    }

    pub async fn write_credentials(&self, credentials: &StoredCredentials) -> Result<()> {
        if self.keyring.write_credentials(credentials)? {
            return Ok(());
        }

        let key = self.get_or_create_encryption_key().await?;
        let payload = encrypt_string(&serde_json::to_string(credentials)?, &key)?;
        self.write_json_file(&self.credentials_path, &payload).await
    }

    async fn ensure_base_dir(&self) -> Result<()> {
        fs::create_dir_all(&self.base_dir)
            .await
            .with_context(|| format!("Failed to create {}", self.base_dir.display()))?;
        set_mode(&self.base_dir, 0o700).await
    }

    async fn read_json_file<T>(&self, file_path: &Path) -> Result<Option<T>>
    where
        T: DeserializeOwned,
    {
        match fs::read_to_string(file_path).await {
            Ok(contents) => Ok(Some(
                serde_json::from_str(&contents)
                    .with_context(|| format!("Failed to parse {}", file_path.display()))?,
            )),
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(None),
            Err(error) => {
                Err(error).with_context(|| format!("Failed to read {}", file_path.display()))
            }
        }
    }

    async fn write_json_file<T>(&self, file_path: &Path, value: &T) -> Result<()>
    where
        T: Serialize,
    {
        self.ensure_base_dir().await?;

        let temp_path = file_path.with_extension(format!(
            "{}.tmp",
            file_path
                .extension()
                .and_then(|value| value.to_str())
                .unwrap_or("")
        ));
        let contents = format!(
            "{}\n",
            serde_json::to_string_pretty(value)
                .with_context(|| format!("Failed to serialize {}", file_path.display()))?
        );

        fs::write(&temp_path, contents)
            .await
            .with_context(|| format!("Failed to write {}", temp_path.display()))?;
        set_mode(&temp_path, 0o600).await?;
        fs::rename(&temp_path, file_path)
            .await
            .with_context(|| format!("Failed to move {} into place", temp_path.display()))?;
        set_mode(file_path, 0o600).await
    }

    async fn get_or_create_encryption_key(&self) -> Result<Vec<u8>> {
        self.ensure_base_dir().await?;

        match fs::read_to_string(&self.key_path).await {
            Ok(value) => STANDARD
                .decode(value.trim())
                .context("Failed to decode stored encryption key"),
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
                let mut key = vec![0u8; 32];
                OsRng.fill_bytes(&mut key);
                fs::write(&self.key_path, STANDARD.encode(&key))
                    .await
                    .with_context(|| format!("Failed to write {}", self.key_path.display()))?;
                set_mode(&self.key_path, 0o600).await?;
                Ok(key)
            }
            Err(error) => {
                Err(error).with_context(|| format!("Failed to read {}", self.key_path.display()))
            }
        }
    }
}

fn encrypt_string(plaintext: &str, key: &[u8]) -> Result<EncryptedPayload> {
    let cipher = Aes256Gcm::new_from_slice(key)
        .map_err(|_| anyhow::anyhow!("Failed to initialize AES-256-GCM"))?;
    let mut iv = [0u8; 12];
    OsRng.fill_bytes(&mut iv);
    let ciphertext_with_tag = cipher
        .encrypt(Nonce::from_slice(&iv), plaintext.as_bytes())
        .map_err(|_| anyhow::anyhow!("Failed to encrypt credentials"))?;
    let split_index = ciphertext_with_tag
        .len()
        .checked_sub(16)
        .context("Encrypted payload was shorter than the GCM tag")?;

    let (ciphertext, tag) = ciphertext_with_tag.split_at(split_index);

    Ok(EncryptedPayload {
        iv: STANDARD.encode(iv),
        tag: STANDARD.encode(tag),
        ciphertext: STANDARD.encode(ciphertext),
    })
}

fn decrypt_string(payload: &EncryptedPayload, key: &[u8]) -> Result<String> {
    let cipher = Aes256Gcm::new_from_slice(key)
        .map_err(|_| anyhow::anyhow!("Failed to initialize AES-256-GCM"))?;
    let iv = STANDARD
        .decode(&payload.iv)
        .context("Failed to decode IV")?;
    let tag = STANDARD
        .decode(&payload.tag)
        .context("Failed to decode tag")?;
    let ciphertext = STANDARD
        .decode(&payload.ciphertext)
        .context("Failed to decode ciphertext")?;

    let mut combined = ciphertext;
    combined.extend_from_slice(&tag);
    let plaintext = cipher
        .decrypt(Nonce::from_slice(&iv), combined.as_ref())
        .map_err(|_| anyhow::anyhow!("Failed to decrypt credentials"))?;

    String::from_utf8(plaintext).context("Decrypted credentials were not valid UTF-8")
}

#[cfg(unix)]
async fn set_mode(path: &Path, mode: u32) -> Result<()> {
    use std::os::unix::fs::PermissionsExt;

    let permissions = std::fs::Permissions::from_mode(mode);
    fs::set_permissions(path, permissions)
        .await
        .with_context(|| format!("Failed to set permissions on {}", path.display()))
}

#[cfg(not(unix))]
async fn set_mode(_path: &Path, _mode: u32) -> Result<()> {
    Ok(())
}
